mesa: fix cross compilation to x86_64

alyssais opened the pull request:
###### Description of changes

In particular, fix cross compilation of rusticl.  Previously, Nixpkgs has not been equipped to handle non-Cargo Rust cross compilation.  That's getting increasingly important, now that we're building Mesa with Rust, and soon the kernel, and [potentially](https://github.com/systemd/systemd/issues/26640) systemd in not too long.

This is a bigger PR than I usually like to make, but it's difficult to explain each of these changes out of context, so I've bundled them into a single PR.  I'd definitely recommend reviewing commit by commit though.  The commit messages contain the details of each change.

Spiritual successor to #159064.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
lib.systems: elaborate Rust metadata

We need this stuff to be available in lib so make-derivation.nix can
access it to construct the Meson cross file.

This has a couple of other advantages:

 - It makes Rust less special.  Now figuring out what Rust calls a
   platform is the same as figuring out what Linux or QEMU call it.

 - We can unify the schema used to define Rust targets, and the schema
   used to access those values later.  Just like you can set "config"
   or "system" in a platform definition, and then access those same
   keys on the elaborated platform, you can now set "rustcTarget" in
   your crossSystem, and then access "stdenv.hostPlatform.rustcTarget"
   in your code.

"rustcTarget", "rustcTargetSpec", "cargoShortTarget", and
"cargoEnvVarTarget" have the "rustc" and "cargo" prefixes because
these are not exposed to code by the compiler, and are not
standardized.  The arch/os/etc. variables are all named to match the
forms in the Rust target spec JSON.

The new rust.target-family only takes a list, since we don't need to
worry about backwards compatibility when that name is used.

The old APIs are all still functional with no warning for now, so that
it's possible for external code to use a single API on both 23.05 and
23.11.  We can introduce the warnings once 23.05 is EOL, and make them
hard errors when 23.11 is EOL.

alyssais committed:
stdenv: fix meson rust cross compilation

alyssais committed:
mesa: fix cross compilation to x86_64

In particular, fix cross compilation of rusticl.

RaitoBezarius commented:
Result of `nixpkgs-review pr 233628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>mesa</li>
    <li>mesa.debug (mesa.debug.debug ,mesa.debug.dev ,mesa.debug.drivers ,mesa.debug.driversdev ,mesa.debug.opencl ,mesa.debug.osmesa ,mesa.debug.spirv2dxil)</li>
    <li>mesa.dev (mesa.dev.debug ,mesa.dev.dev ,mesa.dev.drivers ,mesa.dev.driversdev ,mesa.dev.opencl ,mesa.dev.osmesa ,mesa.dev.spirv2dxil)</li>
    <li>mesa.drivers (mesa.drivers.debug ,mesa.drivers.dev ,mesa.drivers.drivers ,mesa.drivers.driversdev ,mesa.drivers.opencl ,mesa.drivers.osmesa ,mesa.drivers.spirv2dxil)</li>
    <li>mesa.driversdev (mesa.driversdev.debug ,mesa.driversdev.dev ,mesa.driversdev.drivers ,mesa.driversdev.driversdev ,mesa.driversdev.opencl ,mesa.driversdev.osmesa ,mesa.driversdev.spirv2dxil)</li>
    <li>mesa.opencl (mesa.opencl.debug ,mesa.opencl.dev ,mesa.opencl.drivers ,mesa.opencl.driversdev ,mesa.opencl.opencl ,mesa.opencl.osmesa ,mesa.opencl.spirv2dxil)</li>
    <li>mesa.osmesa (mesa.osmesa.debug ,mesa.osmesa.dev ,mesa.osmesa.drivers ,mesa.osmesa.driversdev ,mesa.osmesa.opencl ,mesa.osmesa.osmesa ,mesa.osmesa.spirv2dxil)</li>
    <li>mesa.spirv2dxil (mesa.spirv2dxil.debug ,mesa.spirv2dxil.dev ,mesa.spirv2dxil.drivers ,mesa.spirv2dxil.driversdev ,mesa.spirv2dxil.opencl ,mesa.spirv2dxil.osmesa ,mesa.spirv2dxil.spirv2dxil)</li>
  </ul>
</details>

RaitoBezarius commented:
Result of `nixpkgs-review pr 233628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>mesa</li>
    <li>mesa.debug (mesa.debug.debug ,mesa.debug.dev ,mesa.debug.drivers ,mesa.debug.driversdev ,mesa.debug.opencl ,mesa.debug.osmesa ,mesa.debug.spirv2dxil)</li>
    <li>mesa.dev (mesa.dev.debug ,mesa.dev.dev ,mesa.dev.drivers ,mesa.dev.driversdev ,mesa.dev.opencl ,mesa.dev.osmesa ,mesa.dev.spirv2dxil)</li>
    <li>mesa.drivers (mesa.drivers.debug ,mesa.drivers.dev ,mesa.drivers.drivers ,mesa.drivers.driversdev ,mesa.drivers.opencl ,mesa.drivers.osmesa ,mesa.drivers.spirv2dxil)</li>
    <li>mesa.driversdev (mesa.driversdev.debug ,mesa.driversdev.dev ,mesa.driversdev.drivers ,mesa.driversdev.driversdev ,mesa.driversdev.opencl ,mesa.driversdev.osmesa ,mesa.driversdev.spirv2dxil)</li>
    <li>mesa.opencl (mesa.opencl.debug ,mesa.opencl.dev ,mesa.opencl.drivers ,mesa.opencl.driversdev ,mesa.opencl.opencl ,mesa.opencl.osmesa ,mesa.opencl.spirv2dxil)</li>
    <li>mesa.osmesa (mesa.osmesa.debug ,mesa.osmesa.dev ,mesa.osmesa.drivers ,mesa.osmesa.driversdev ,mesa.osmesa.opencl ,mesa.osmesa.osmesa ,mesa.osmesa.spirv2dxil)</li>
    <li>mesa.spirv2dxil (mesa.spirv2dxil.debug ,mesa.spirv2dxil.dev ,mesa.spirv2dxil.drivers ,mesa.spirv2dxil.driversdev ,mesa.spirv2dxil.opencl ,mesa.spirv2dxil.osmesa ,mesa.spirv2dxil.spirv2dxil)</li>
  </ul>
</details>

vcunat commented:
I looked just at the mesa diff and there I see no potential issue.  So `glslang` is really needed as both native and non-native?

alyssais commented:
> So `glslang` is really needed as both native and non-native?

Yes.  Build dep on glslangValidator, host dep on SPIRV-Tools.

RaitoBezarius commented:
Confirmed `/nix/store/h5kraiprihi263lbsl19n6ssy1bl8ynp-mesa-x86_64-unknown-linux-gnu-23.0.3` built from i686-linux to x86_64-linux (I didn't have fast aarch64-linux but I can setup a M1 if this is needed).

RaitoBezarius reviewed (approved):
Tested mesa, cross compilation from i686 to x86_64 for mesa and diff LGTM to me.

Ericson2314 commented:
I have some doubts on the first commit --- @alyssais and I are talking and I may yet be persuaded :) --- but the other 3 commits look great!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-nixbsd-a-posibility/29612/3

Ericson2314 commented:
Wait I thought I relented on the first commit and this was done months ago?

alyssais commented:
> Wait I thought I relented on the first commit and this was done months ago?

Nope, I agreed that it should be unified with the elaboration params and then just took forever to find the time for another round.

alyssais commented on lib/systems/default.nix:
```diff
@@ -121,10 +121,89 @@ rec {
+        # https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch
+        arch =
+          /**/ if args ? rust.arch then args.rust.arch
+          # TODO: deprecate rustc.platform in favour of rust.rustcPlatform after 23.05 is EOL.
```
Fixed.  There's now "rustc.platform" (old, to be deprecated — as you pointed out on Matrix a while ago, this is a bad name), and "rust.platform" (new).

alyssais commented:
Rebased. Dropped the bindgen wrapper, since Meson now supports specifying bindgen clang arguments.  rust.lib is no longer completely obsolete, due to the addition of rust.lib.envVars in the meantime, which wouldn't make sense to move to lib.

alyssais commented on lib/systems/default.nix:
```diff
@@ -159,10 +159,112 @@ rec {
         } // platforms.select final)
         linux-kernel gcc rustc;
 
+      rust = {
```
Done.

amjoseph-nixpkgs commented:
I'm definitely in favor of the first commit.  If you want to put it in a separate PR I would approve that as soon as I finish rebuilding with it.  Or I can just comment here when that finishes.

The other two commits LGTM but I don't know meson very well and I'm going to have my mesa pinned back to `22.3.7` for a few more weeks (long story).

alyssais commented:
> I'm definitely in favor of the first commit. If you want to put it in a separate PR I would approve that as soon as I finish rebuilding with it.

Thanks for having a look. :)

I'm still waiting to make sure that @Ericson2314 is okay with the first commit (and AIUI he's definitely okay with the rest of them already), so I don't think splitting out the first one would speed up getting it merged at all.

alyssais commented:
I've rebased to fix the merge conflict, and I've [also](https://github.com/NixOS/nixpkgs/compare/576274227287a3d630245406746bcfb3a1f07c68..c0d3cbb3592694d5062b68909dc38d3a0ffd426a):

- Simplified the logic for looking up values from `args.rust` falling back to `args.rustc` falling back to a default.
- Added an assert so that it's not possible to pass `args.rust` (the new name) and `args.rustc` (the old name) at the same time.
- Fixed `rustcTargetSpec` referring to a non-existent `rustcPlatform` that was left over from a previous iteration.

All suggested by @Ericson2314.

amjoseph-nixpkgs commented on lib/systems/default.nix:
```diff
@@ -159,9 +163,101 @@ rec {
+        # True if the target is no_std
+        # https://github.com/rust-lang/rust/blob/2e44c17c12cec45b6a682b1e53a04ac5b5fcc9d2/src/bootstrap/config.rs#L415-L421
+        isNoStdTarget =
+          builtins.any (t: lib.hasInfix t final.rust.rustcTarget) ["-none" "nvptx" "switch" "-uefi"];
```
I'd really like to not use string-pattern-matching for this...

```suggestion
          final.isNone;
```

`isUefi` is added by https://github.com/NixOS/nixpkgs/pull/231951 so once that merges we can append a `|| final.isUefi` here.

What is `switch`?

Anybody targeting `nvptx` is going to have to construct their own platform object anyways; they can just set `isNoStdTarget`.

alyssais replied:
I'd like that too, but I don't want to change how it works in this PR, because it's off-topic.

amjoseph-nixpkgs commented on pkgs/build-support/rust/lib/default.nix:
```diff
@@ -5,89 +5,6 @@
 
 rec {
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch
-  toTargetArch = platform:
-    /**/ if platform ? rustc.platform then platform.rustc.platform.arch
-    else if platform.isAarch32 then "arm"
-    else if platform.isMips64  then "mips64"     # never add "el" suffix
-    else if platform.isPower64 then "powerpc64"  # never add "le" suffix
-    else platform.parsed.cpu.name;
-
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_os
-  toTargetOs = platform:
-    /**/ if platform ? rustc.platform then platform.rustc.platform.os or "none"
-    else if platform.isDarwin then "macos"
-    else platform.parsed.kernel.name;
-
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_family
-  toTargetFamily = platform:
```
We can't delete these without a deprecation period.

The latest release of `crate2nix` emits code that uses them.

Could you please add `lib.warn` placeholders?  We can remove them after the branch-off.

alyssais replied:
They're not being deleted.  They're at the bottom of the file.

amjoseph-nixpkgs replied:
```
       error: attribute 'arch' missing

       at /git/ubik/nixpkgs/pkgs/build-support/rust/lib/default.nix:70:5:

           69|   # lib.warn "`rust.${old} platform` is deprecated. Use `platform.rust.${new}` instead."
           70|     platform.rust.${new})
             |     ^
           71| {
```

amjoseph-nixpkgs replied:
> They're at the bottom of the file.

All I see at the bottom of the file is

```nix
  toTargetArch = "arch";
```

but this doesn't work (see error above; this happens for any `Cargo.nix` produced by `crate2nix`).

It might be the case that `crate2nix` is applying `toTargetArch` to things you didn't expect it to.  It should stop doing that, but we should maintain the old behavior for one cycle with a `lib.warn` in order to let that migration happen.

amjoseph-nixpkgs replied:
Also, I'm totally in favor of the refactoring in d51b7f8893470d998a344acfc2eefdbdf413b2cd, but it seems a little bit weird to hide it in a PR whose title is mostly about cross-compiling a C/C++ codebase (mesa).  Could you just expose whatever extra bits you need exposed for meson, and then do the cleanup in a separate PR?  I don't want waiting for `crate2nix` to switch to the new API to block `mesa` from working...

amjoseph-nixpkgs replied:
If I revert the first and last chunks changed in `pkgs/build-support/rust/lib/default.nix` everything works again.

Could we split just those two chunks out into a separate PR in order to unblock things?  Yes I know that means there will be duplicate code.  But it will maintain the existing behavior.

alyssais replied:
The refactoring is necessary so that we can access this data in make-derivation.nix, which is necessary to fix Mesa cross compilation.  I've been working on fixing this problem with Mesa for a year and a half.  I don't think another few days while last-minute bugs are going to make a difference.  I definitely don't want to have duplicate code, because it might start diverging, and what if something comes up and I don't have the capacity to clean it up straight away?  It's much better to get it fixed atomically.

alyssais replied:
Fixed.  It had just got out of sync with all the different revisions this has gone through, and had to be updated to use the correct new attributes.

amjoseph-nixpkgs reviewed (changes requested):
We need `lib.warn` backwards-compatibility placeholders on the `toFooBar` attributes; `crate2nix` uses them (in their current location).

Otherwise looks good.

Building:
- [ ] pkgs.tests.cross.sanity

alyssais commented:
> We need `lib.warn` backwards-compatibility placeholders on the `toFooBar` attributes; `crate2nix` uses them (in their current location).

There are warnless aliases in rust/lib/default.nix already?

amjoseph-nixpkgs reviewed (approved):
Rust changes look great.

Built:

- [x] `tests.cross.sanity`
- :skull_and_crossbones: `pkgsCross.gnu64_simplekernel.mesa-23.1.9`
  - Needs https://github.com/NixOS/nixpkgs/pull/265891
  - Needs https://github.com/NixOS/nixpkgs/pull/265162
  - Probably needs https://github.com/NixOS/nixpkgs/pull/240435

Mic92 reviewed (approved):
Looks clean.

Ericson2314 commented on lib/systems/default.nix:
```diff
@@ -159,9 +163,101 @@ rec {
+        platform = rust.platform or {} // {
+          # https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch
+          arch =
+            /**/ if rust ? platform.arch then rust.platform.arch
```
```suggestion
            /**/ if rust ? platform then rust.platform.arch
```

Ericson2314 replied:
It is a required field of `platform`

alyssais replied:
Right, yeah, I see now that this was how it worked before.

alyssais replied:
Fixed.

Ericson2314 reviewed (approved):
Just one little thing, and then it looks good to. This is very important change let's get it in ~~23.05~~ 23.11!

(I didn't really look at all the usage site, but I think those are very low risk)

amjoseph-nixpkgs commented:
> This is very important change let's get it in 23.05!

I think that ship has sailed :laughing:

alyssais commented:
The sphinx build failure on x86_64-linux seems to be intermittent / load-related.  Let's try again:

@OfBorg build mesa mesa.passthru.tests stdenv stdenv.passthru.tests

amjoseph-nixpkgs commented:
> The sphinx build failure on x86_64-linux seems to be intermittent / load-related. Let's try again:

I've been getting `sphinx` test failures on my machines too.  It's annoying.  I disabled the tests for that package.  This has happened before btw.

alyssais commented:
OfBorg hasn't given me the aarch64 builds for days, so I'm just going to merge.  I've tested locally on x86_64-linux and aarch64-linux, and IIRC mesa is broken on Darwin anyway.  In the unlikely event there are failures, we'll catch them in staging.

andresilva commented:
Seems like this PR broke all packages that use `rustc-wasm32`, namely: `lldap`, `pagefind` and `polkadot`.

andresilva commented:
I found https://github.com/NixOS/nixpkgs/pull/267876 which should fix this.

alyssais commented:
Fix is https://github.com/NixOS/nixpkgs/pull/268168.
